[WIP] DO NOT MERGE: Distribute docker images and allow examples to use docker

dora-gt opened the pull request:
Fixes #179 Distribute pre-built binaries and/or allow examples to use Docker

**This should be merged after #236 is merged**

- Docker files are provided (`node.dockerfile`, `eth-se.dockerfile`)
    - based on `Dockerfile`
- Docker images will be available on [Docker Hub](https://cloud.docker.com/u/interledgerrs/repository/list)
    - Have to turn on [Docker Hub builds](https://docs.docker.com/docker-hub/builds/) for both `.dockerfile`
- Examples use `docker` command to run `node` or `ethereum-ledger`
- Don't use `docker-compose` because these are examples and the main objective is to train readers. If we use `docker-compose`, it is too easy to tell the **real** procedure and the settings.

Questions:
- Should we automatically obtain XRP ledger private key/address pair from [XRP Testnet Faucet](https://xrpl.org/xrp-testnet-faucet.html)? because it seems that the testnet addresses are reset every a few months (then examples will get broken)
    - That said, this could be detectable if we write `README` tests.
- Should `node.dockerfile` be named `Dockerfile`?

emschwartz commented on eth-se.dockerfile:
```diff
@@ -0,0 +1,41 @@
+# You can do this by adding the option `-v data-volume-name:/data`
+# when calling `docker run`.
+
+VOLUME [ "/data" ]
```
Doesn't this also need to connect to a Redis instance? (Or are we going to connect it to an instance that lives outside the container? If so, I'm not sure we need this volume)

gakonst replied:
I believe we should package it with Redis (at least for this release).

dora-gt replied:
Sorry, I just forgot to remove this. We don't need this because Redis is on another container.

@gakonst, What do you mean by package? Do you mean squashing Redis into SE's container?
Is your image like this?

- Node Container
    - Interledger Node
    - Redis for Interledger Node
- ETH SE Container
    - ETH SE
    - Redis for ETH SE

gakonst replied:
I'd prefer it like that I think. In the future, that would mean we could have the Interledger node and the SE running a different type fo database. For now the current approach might be OK

dora-gt replied:
I see. I intended that these `.dockerfile` s are used for examples and for normal uses, so in the latter case, I'm not sure we should package it with Redis because it could be SQL or something.

So what about mentioning it in the document.

dora-gt replied:
I mentioned like this:
https://github.com/interledger-rs/interledger-rs/pull/270/files#diff-f32e61e560deb8bb8ef2b67caf3dfc1dR75

emschwartz commented on examples/eth-settlement/README.md:
```diff
@@ -8,169 +8,201 @@ This example shows how to configure Interledger.rs nodes and use an Ethereum net
-- [Redis](#redis)
 
-### Rust
+### Docker
```
I think we should try to make it so the examples can either be run with Docker or without. Some people may want to get more hands-on with the Rust code and would prefer not to interact with it through Docker

dora-gt replied:
Hmm yes, indeed. I'm not sure I can cleanly realize it but anyway I'll try.

dora-gt replied:
I think there are 3 options,

1. Insert both sub-sections.
    1. for non-Docker users
    1. for Docker users
1. Provide both sections
    1. for non-Docker users
    1. for Docker users
1. Provide separate documents

Options 2 and 3 contain a lot of duplicated areas, so I'll try option 1 first. If that looks unexpectedly hard to read, I'll consider option 2 or 3.

dora-gt replied:
hmmmmm, Option 1 looks not so good... this is confusing and hard to read.

I feel like option 3. Although this could increase the cost of maintenance, I think this is better for readers. If we prioritize DX, we should adopt option 3. I'll try option 3.

![image](https://user-images.githubusercontent.com/811594/64312091-8c4da200-cfe1-11e9-8e1b-377e484477c1.png)

dora-gt replied:
Now the directory structure looks like:
```
├── README.md
├── eth-settlement
│   ├── README.md
│   ├── procedure-docker.md
│   └── procedure.md
├── eth_xrp_three_nodes
│   ├── README.md
│   ├── procedure-docker.md
│   └── procedure.md
├── simple
│   ├── README.md
│   ├── procedure-docker.md
│   └── procedure.md
```
Each example `procedure.md` which is for non docker users, and `procedure-docker.md` which is for docker users. These 2 are linked from `README.md`s.

emschwartz replied:
Do you think it would be possible to have one instruction file and use a bash variable to store the run commands for the node and SE?

The idea would be that if you run `run-md.sh README.md` it would use `cargo run` by default but if you run `USE_DOCKER=1 run-md.sh README.md` it would use the appropriate `docker run...` command instead. I'm imagining that the script would set a local variable like `RUN_NODE` to be either `cargo run...` or `docker run...` depending on that environment variable.

If we could make this work, it seems like this would require the least amount of duplication

dora-gt replied:
Ah, that is what I tried first. The `.md` of the image(https://github.com/interledger-rs/interledger-rs/pull/270#discussion_r321070479) was like that.

I think it is possible but even if we do that, the explanations become too complex because there are many sub-sections.

Or... do you mean just changing command and leave the explanations for non-docker version?

emschwartz replied:
I think it would be fine to only have the non-docker explanation for now. If people find it confusing, we could change it but we don't really want to spend time explaining docker to people that don't know it. We just want to explain how to use the Interledger components and the instructions should be very similar with the exception of docker-specific things like setting up the networking

dora-gt replied:
I see. Explaining docker might be too much. I think we need to place prerequisites on explanations, which means deciding what we should explain and what we won't try to explain. If we define `docker` as prerequisites, that would be fine.

I'll try.

emschwartz commented on examples/eth-settlement/README.md:
```diff
@@ -8,169 +8,201 @@ This example shows how to configure Interledger.rs nodes and use an Ethereum net
+NETWORK_ID=`docker network ls -f "name=interledger" --format="{{.ID}}"`
+if [ -z "${NETWORK_ID}" ]; then
+    printf "Creating a docker network...\n"
+    docker network create interledger
```
Won't this be called twice?

dora-gt replied:
On my environment, it won't. Why do you think so?

emschwartz commented on examples/eth-settlement/README.md:
```diff
@@ -1,491 +1,23 @@
-The `settle_threshold` and `settle_to` parameters control when settlements are triggered. The node will send a settlement when an account's balance reaches the `settle_threshold`, and it will settle for `balance - settle_to`.
-
-### 7. Sending a Payment
+- [Compile from the source code and deploy](./procedure.md)⚙
```
If we offer both, we should make one of them the default so that you only need to click over to a different file if you're not using the default method

emschwartz replied:
I would suggest that the non-docker method should be the default

dora-gt replied:
Although it depends on the decision of https://github.com/interledger-rs/interledger-rs/pull/270#discussion_r321450828, what about this structure?

```
├── simple
│   ├── README.md (default procedure, add link to docker.md)
│   └── docker.md
```

dora-gt replied:
(docker version will be merged into non docker version without explanations)

emschwartz commented on examples/eth-settlement/images/overview-docker.svg:
```diff
@@ -0,0 +1,202 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
It would be helpful to show what is in each docker container, since there are multiple containers running

dora-gt replied:
I agree but actually these are almost 1:1. So every box indicates that it is a container. So I think mentioning it on documents would be sufficient. Placing some explanation in the diagram makes it complicated, I think.

![image](https://user-images.githubusercontent.com/811594/64434740-33266100-d0fc-11e9-91ac-8630368ca15a.png)

emschwartz replied:
I see, then maybe it's just the box around it that's confusing because it seems to suggest that everything is running in a single docker container rather than everything as a separate docker container

dora-gt replied:
(Resolve this because we no longer need this diagram https://github.com/interledger-rs/interledger-rs/pull/270#discussion_r322049158)

emschwartz commented on examples/eth-settlement/procedure-docker.md:
```diff
@@ -0,0 +1,492 @@
+
+printf "Stopping Interledger nodes\n"
+
+REDIS_CONTAINER_ID=`docker ps -q -f "name=redis"`
```
We should just have each node and SE use a different Redis instance

dora-gt replied:
Then isn't it natural that `Simple` example has 2 redis, one for Node A and another for Node B? Should I change that too?

dora-gt replied:
I change so for now, to make it consistent with the other examples.

emschwartz replied:
I agree with that

dora-gt commented:
I think I've addressed most points including Ben's suggestions (#287).
I'm checking if there are any points that additionally I should change.

I'm not sure why I need this:
https://github.com/interledger-rs/interledger-rs/pull/270/files#diff-3301fa36a91ba4f24d42a74c47b0c213R43-R50
but it doesn't work with a pipe. 🤔

emschwartz commented on examples/eth_xrp_three_nodes/README.md:
```diff
@@ -177,33 +249,100 @@ In this example, we'll connect 3 Interledger nodes and each node needs its own s
+printf "\nStarting settlement engines...\n"
+if [ "$USE_DOCKER" -eq 1 ]; then
+    # Start Alice's settlement engine (ETH)
+    docker start interledger-rs-se_a ||
```
nit: this way of starting the nodes prints a lot of errors in the terminal such as the following. It would be great to make sure it doesn't print that:

```
Error response from daemon: No such container: redis-alice_node
Error: failed to start containers: redis-alice_node
```

dora-gt replied:
Yeah... it looks not so nice but I hesitated to suppress the outputs because sometimes it fails because of the other reasons than `No such container`.

I'll try to figure out a better way.

dora-gt replied:
I realized that we didn't need to do this because the containers are always cleaned. Now it must look better.

emschwartz commented on node.dockerfile:
```diff
@@ -0,0 +1,39 @@
+# Build Interledger node into standalone binary
```
Also, if we do have a `docker` directory, can you also move the `redis.conf` and `run-node-and-redis.sh` files into there too?

dora-gt replied:
> Should we move both of these to a docker folder?

Actually I thought it would be better because currently the top directory looks a bit messy.
If you don't mind, I would like to do so.

![image](https://user-images.githubusercontent.com/811594/64586934-b5eb3c80-d3d8-11e9-91ef-b4a5cbbf1e7b.png)

We can flexibly automate builds. That should be fine.

I'm wondering if having these 2 dockerfiles is confusing.

- Docker
- node.dockerfile

I'm not sure we want a packaged image like the current `Dockerfile`.

bstrie commented:
@emschwartz Was this closed by accident?

dora-gt commented:
It seems to be an accident but I can't reopen this (Idk why), so I'll open another PR later.

dora-gt commented:
It seems to be an accident but I can't reopen this (maybe because `feature/cli-improvement/main` branch was deleted as the branch was merged into the master), so I'll open another PR later.

dora-gt commented on scripts/run-md.sh:
```diff
@@ -9,6 +9,8 @@
 # Markdown file but run them when the whole file is executed.
 # This is largely intended for commands that make the output more readable
 
+TMP_SCRIPT=$(mktemp -t "md.sh")
```
Oops, I didn't notice because it works on my environment.

dora-gt replied:
Fixed. (but I also try to figure out why a pipe doesn't work)

dora-gt commented on examples/eth-settlement/README.md:
```diff
@@ -36,7 +36,7 @@ Advanced: You can run this against the Rinkeby Testnet by running a node that co
 
 ### Redis
-The Interledger.rs nodes currently use [Redis](https://redis.io/) to store their data (SQL database support coming soon!)
+The Interledger.rs nodes and settlement engines currently use [Redis](https://redis.io/) to store their data (SQL database support coming soon!). Nodes and settlement engines don't need to share the Redis necessarily.
```
Fixed

dora-gt commented on examples/eth-settlement/README.md:
```diff
@@ -49,106 +49,277 @@ Make sure your Redis is empty. You could run `redis-cli flushall` to clear all t
+if [ "$USE_DOCKER" -eq 1 ]; then
+    docker --version > /dev/null || error_and_exit "Uh oh! You need to install Docker before running this example"
+    
+    docker stop \
```
Fixed.

dora-gt commented on examples/eth-settlement/README.md:
```diff
@@ -158,46 +329,116 @@ ILP_REDIS_URL=redis://127.0.0.1:6379/0 \
 ILP_SECRET_SEED=1604966725982139900555208458637022875563691455429373719368053354 \
 ILP_ADMIN_AUTH_TOKEN=hi_bob \
-ILP_REDIS_URL=redis://127.0.0.1:6379/1 \
+ILP_REDIS_URL=redis://127.0.0.1:6381/0 \
```
Fixed.

dora-gt commented:
New PR is here: #293
